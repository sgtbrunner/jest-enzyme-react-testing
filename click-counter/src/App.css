.app {
  text-align: center;
}

.error {
  color: red;
  margin-top: 20px;
}

.hidden {
  display: none;
}
