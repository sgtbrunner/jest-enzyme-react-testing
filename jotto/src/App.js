import React, { useEffect } from "react";

import Congrats from "./Congrats";
import Input from "./Input";
import GuessedWords from "./GuessedWords";
import { getSecretWord } from "./actions";

function App() {
  // TODO: get props from shared state
  const success = false;
  const secretWord = "party";
  const guessedWords = [];

  useEffect(() => {
    getSecretWord();
  }, []);

  return (
    <div className="container" data-test-id="app-component">
      <h1>Jotto</h1>
      <Congrats success={success} />
      <Input success={success} secretWord={secretWord} />
      <GuessedWords guessedWords={guessedWords} />
    </div>
  );
}

export default App;
