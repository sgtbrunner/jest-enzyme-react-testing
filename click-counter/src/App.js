import React, { useState } from "react";
import "./App.css";

const initialCount = 0;

const App = () => {
  const [count, setCount] = useState(initialCount);
  const [error, setError] = useState(false);

  const decrementCounter = () => {
    if (count > 0) setCount(count - 1);
    else setError(true);
  };

  const incrementCounter = () => {
    setCount(count + 1);
    if (error) setError(false);
  };

  const resetCounter = () => {
    setCount(initialCount);
    if (error) setError(false)
  }

  return (
    <div data-test-id="component-app" className="app">
      <h1 data-test-id="counter-display">
        The counter is currently&nbsp;
        <span data-test-id="counter">{count}</span>
      </h1>
      <button data-test-id="increment-button" onClick={incrementCounter}>
        Increment counter
      </button>
      <button data-test-id="decrement-button" onClick={decrementCounter}>
        Decrement counter
      </button>
      <button
        data-test-id="reset-button"
        onClick={resetCounter}
      >
        Reset counter
      </button>
      {error && (
        <p data-test-id="error-message" className="error">
          The counter cannot go below 0
        </p>
      )}
    </div>
  );
};

export default App;
