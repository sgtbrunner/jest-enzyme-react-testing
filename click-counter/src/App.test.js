import Enzyme, { shallow } from "enzyme";
import EnzymeAdapter from "@wojtekmaj/enzyme-adapter-react-17";
import App from "./App";

Enzyme.configure({ adapter: new EnzymeAdapter() });

const setup = () => shallow(<App />);

const findByTestAttr = (wrapper, value) =>
  wrapper.find(`[data-test-id='${value}']`);

const COUNTER = "counter";
const INCREMENT_BUTTON = "increment-button";
const DECREMENT_BUTTON = "decrement-button";
const RESET_BUTTON = "reset-button";
const ERROR_MESSAGE = "error-message";
const HIDDEN = "hidden";
const CLICK = "click";

let wrapper;

describe("App component rendering", () => {
  beforeEach(() => {
    wrapper = setup();
  });

  it("renders App without crashing", () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders without error", () => {
    const appComponent = findByTestAttr(wrapper, "component-app");
    expect(appComponent.length).toBe(1);
  });

  it("renders counter display", () => {
    const counterDisplay = findByTestAttr(wrapper, "counter-display");
    expect(counterDisplay.length).toBe(1);
  });

  it("counter starts at 0", () => {
    const count = findByTestAttr(wrapper, COUNTER).text();
    expect(count).toEqual("0");
  });

  it("renders increment button", () => {
    const incrementButton = findByTestAttr(wrapper, INCREMENT_BUTTON);
    expect(incrementButton.length).toBe(1);
  });

  it("renders decrement button", () => {
    const decrementButton = findByTestAttr(wrapper, DECREMENT_BUTTON);
    expect(decrementButton.length).toBe(1);
  });

  it("renders reset button", () => {
    const resetButton = findByTestAttr(wrapper, RESET_BUTTON);
    expect(resetButton.length).toBe(1);
  });
});

describe("Increment button action", () => {
  it("clicking increment button increments counter accordingly", () => {
    const incrementButton = findByTestAttr(wrapper, INCREMENT_BUTTON);
    incrementButton.simulate(CLICK);

    const count = findByTestAttr(wrapper, COUNTER).text();
    expect(count).toEqual("1");
  });
});

describe("Decrement button action", () => {
  it(`clicking decrement button decrements counter accordingly if counter
    is greater than 0`, () => {
    const wrapper = setup();
    const incrementButton = findByTestAttr(wrapper, INCREMENT_BUTTON);
    incrementButton.simulate(CLICK);

    const decrementButton = findByTestAttr(wrapper, DECREMENT_BUTTON);
    decrementButton.simulate(CLICK);

    const count = findByTestAttr(wrapper, COUNTER).text();
    expect(count).toEqual("0");
  });
});

describe("Reset button action", () => {
  it(`clicking reset button resets counter accordingly if counter
    is greater than 0`, () => {
    const incrementButton = findByTestAttr(wrapper, INCREMENT_BUTTON);
    incrementButton.simulate(CLICK);

    const resetButton = findByTestAttr(wrapper, RESET_BUTTON);
    resetButton.simulate(CLICK);

    const count = findByTestAttr(wrapper, COUNTER).text();
    expect(count).toEqual("0");
  });
});

describe(`counter is 0 and decrement is clicked`, () => {
  beforeEach(() => {
    const button = findByTestAttr(wrapper, DECREMENT_BUTTON);
    button.simulate(CLICK);
  });

  it(`clicking decrement button decrements doesn't update counter if current
  count is 0`, () => {
    const count = findByTestAttr(wrapper, COUNTER).text();
    expect(count).toEqual("0");
  });

  it("error message shows", () => {
    const error = findByTestAttr(wrapper, ERROR_MESSAGE);
    const errorHasHiddenClass = error.hasClass(HIDDEN);
    expect(errorHasHiddenClass).toBe(false);
  });

  it("clicking increment button clears error", () => {
    const incrementButton = findByTestAttr(wrapper, INCREMENT_BUTTON);
    incrementButton.simulate(CLICK);

    const error = findByTestAttr(wrapper, ERROR_MESSAGE);
    const errorHasHiddenClass = error.hasClass(HIDDEN);
    expect(errorHasHiddenClass).toBe(true);
  });

  it("clicking reset button clears error", () => {
    const resetButton = findByTestAttr(wrapper, RESET_BUTTON);
    resetButton.simulate(CLICK);

    const error = findByTestAttr(wrapper, ERROR_MESSAGE);
    const errorHasHiddenClass = error.hasClass(HIDDEN);
    expect(errorHasHiddenClass).toBe(true);
  });
});
