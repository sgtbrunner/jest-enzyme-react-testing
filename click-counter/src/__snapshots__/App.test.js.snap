// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component rendering renders App without crashing 1`] = `ShallowWrapper {}`;
